src/Haskell/Ide/Engine/Dispatcher.hs

```diff
@@ -43,18 +66,39 @@ mainDispatcher inChan ghcChan ideChan = forever $ do
     Left r ->
       atomically $ writeTChan ideChan r
 
-ideDispatcher :: forall void. DispatcherEnv -> TChan IdeRequest -> IdeM void
-ideDispatcher env pin = forever $ do
+ideDispatcher :: forall void m. DispatcherEnv -> ErrorHandler -> (CallbackHandler m) -> TChan (IdeRequest m) -> IdeM void
+ideDispatcher env errorHandler callbackHandler pin = forever $ do
   debugm "ideDispatcher: top of loop"
   (IdeRequest lid callback action) <- liftIO $ atomically $ readTChan pin
   debugm $ "ideDispatcher:got request with id: " ++ show lid
   cancelled <- liftIO $ atomically $ isCancelled env lid
   unless cancelled $ do
-    res <- action
-    liftIO $ callback res
+    response <- action
+    handleResponse lid callback response
+      
+  where handleResponse lid callback response = case response of
+          IdeResponseResult (IdeResultOk x) -> liftIO $ callbackHandler callback x
+          IdeResponseResult (IdeResultFail err) -> liftIO $ errorHandler lid err 
+          IdeResponseDeferred fp cacheCb -> handleDeferred lid fp cacheCb callback
+    
+        handleDeferred lid fp cacheCb actualCb = queueAction lid fp $ \cm -> do
+          cacheResponse <- cacheCb cm
+          handleResponse lid actualCb cacheResponse
+    
+        queueAction :: J.LspId -> FilePath -> (CachedModule -> IdeM ()) -> IdeM ()
+        queueAction lid fp action = do
+          fp' <- liftIO $ canonicalizePath fp
+          modifyMTState $ \s ->
+            let newReq   = (lid, action)
+                oldQueue = requestQueue s
+                -- add to existing queue if possible
+                newQueue = if Map.member fp' oldQueue
```

Comment:
I am sure this if member stuff can be simplified, by using Map.adjust, or one of the other Map update options